### PR TITLE
Extending the type preservation check

### PR DESCRIPTION
**Pull request by Gaspi**
From the type checking of a LHS of a rule, several kind of constraints may be inferred
1. First-order substitutions: `X = t`
2. Higher-order substitutions: `X x1 ... xn = t`    (under `k`>=`n` abstractions)
where `x1 ... xn` are distinct locally bound variables
3. (Yet) Unprocessed equalities:  `t = u` (possibly under `n` abstractions)
4. Unsatisfiable equalities

The constraints are then used to check convertibility of inferred type of RHS and LHS.
**Expected** behavior for these constraints are:
1. Substitute `X` with `t` in other constraints and in both LHS and RHS expected types
2. Substitute `X a1 ... an` with `t{x1\a1, ..., xn\an}` in other constraints and in both LHS and RHS expected types
*Note : what happens when several rules map the same variable ?*
3. Two possibilities:
    * Record and extend LHS/RHS conversion check to try and substitute in case of exact match (cf Jui-Hsuan internship)
    * Run Knuth-Bendix completion, then locally extend rewriting in the convertibility check (cf Jui-Hsuan internship)
4. The pattern will never match a well-typed term. This is most likely a user error:
-> raise a warning or fail

(1.) and (4.) are already handled in an acceptable way.
This PR should:
- [x] Confirm the SR checking criteria stated above (inputs are welcome)
- [x] Fix issue with (1.) where order of variable is relevant in the substitution (cf `tests/OK/first_order_cstr1.dk`)
- [x] Modify the SR checking algorithm (`typing/pseudo_u` to allow several successive passes of inferring from constraints / process constraints.
- [x] Implement (2.)
- [x] Implement at least the first solution of (3.)
- [x] Write tests for all implemented features

### CODE REVIEW

**Comment by fblanqui** [comment]
Better handling constraints means developing a (better) unification algorithm. Why doing this in Dedukti and not in Lambdapi?
Concerning the use of Knuth-Bendix completion, see http://rewriting.gforge.inria.fr/papers/sr.pdf and https://github.com/Deducteam/lambdapi/pull/200 .

File: commands/dkcheck.ml [hunk 1]
```diff
@@ -60,7 +60,8 @@ let _ =
       c : (confluence) notifies about information provided to the confluence
                      checker (when option --confluence used)
       u : (rule)     provides information about type checking of rules
-      t : (typing)   provides information about type-checking of terms
+      t : (typing)   provides information about type checking of terms
+      s : (SR)       provides information about subject reduction checking of terms
```
Comment (francoisthire) [inline commands/dkcheck.ml]: The SR criterion here is for the rules, not terms in general, right?

File: commands/dkcheck.ml [hunk 2]
```diff
@@ -92,6 +93,11 @@ let _ =
     ; ( "--type-lhs"
       , Arg.Set Typing.fail_on_unsatisfiable_constraints
       , " Forbids rules with untypable left-hand side" )
+    ; ( "--sr-check"
+      , Arg.Int (fun i -> Srcheck.srfuel := i)
+      , "LVL Sets the level of subject reduction checking to LVL.
```
Comment (francoisthire) [inline commands/dkcheck.ml]: LVL Sets... it should be --sr-check right?

File: kernel/exsubst.ml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    - under a given number of lambda abstractions
+    to a term.
+    A substitution raises Not_found meaning that the variable is not subsituted. *)
+type ex_substitution = loc -> ident -> int -> int -> int -> term
+
+(** [apply_exsubst subst n t] applies [subst] to [t] under [n] lambda abstractions.
+      - Variables with DB index [k] <  [n] are considered "locally bound" and are never substituted.
+      - Variables with DB index [k] >= [n] may be substituted if [k-n] is mapped in [sigma]
+          and if they occur applied to enough arguments (substitution's arity). *)
+let apply_exsubst (subst:ex_substitution) (n:int) (te:term) : term*bool =
```
Comment (francoisthire) [inline kernel/exsubst.ml]: Why this function cannot be implemented by one of the subst function? Why this function is not in the module Subst?

Comment (francoisthire) [reply kernel/exsubst.ml]: Ok, I rephrase, why do we have two modules for substitution now? Why don't we keep only one module for substitutions?

**Pull request merged**